### ISSUE TITLE
Incorrect reconciliation of OpenShift Route based on K8s Ingress

### ISSUE THREAD

**Issue by mhofstetter** [openshift/origin#23937]
When OpenShift reconciles and therefore creates an OpenShift `Route` for each Kubernetes `Ingress`, the property `spec.port.targetPort` is set with the name of the port on the `Pod` instead of the name of the port on the referenced `Service`. This gets problematic when they differ in their names, as the referenced name doesn't exist on the `Service`.

Responsible controller: https://github.com/openshift/origin/blob/v3.11.0/pkg/route/controller/ingress/ingress.go#L576

##### Version
OpenShift Master: v3.11.141
Kubernetes Master: v1.11.0+d4cacc0

##### Steps To Reproduce

1. Create a k8s `Service` which targets a named targetPort `default`(on the `Pod`) which differs from the name of the port on the service `http-default`.

```
oc apply -f - << EOF
apiVersion: v1
kind: Service
metadata:
  name: app
spec:
  ports:
  - name: http-default
    port: 8080
    protocol: TCP
    targetPort: default
  selector:
    deployment: app
EOF
```

2. Create a k8s `Ingress` which targets the previously created `Service` on its named service port `http-default`.

```
oc apply -f - << EOF
apiVersion: extensions/v1beta1
kind: Ingress
metadata:
  name: app
spec:
  rules:
  - host: app.test.com
    http:
      paths:
      - backend:
          serviceName: app
          servicePort: http-default
        path: /
EOF
```

##### Current Result

OpenShift creates a `Route` which targets the service `app` but the targetPort is the name of the `Pod` (`default`)  instead of the one of the `Service` (`http-default`).

```
oc get route app-6dqpg -o yaml
apiVersion: route.openshift.io/v1
kind: Route
metadata:
  ...
spec:
  host: app.test.com
  path: /
  port:
    targetPort: default
  to:
    kind: Service
    name: app
...
```

Therefore no connection is possible via the `Route`. Even the OpenShift webconsole warns about this ("Route target port is set to "default", but service "app" has no port with that name.")

##### Expected Result

The created `Route` should target the name of the port on the `Service` instead of the one of the `Pod`.

##### Additional Information

* Relevant sourcecode location: https://github.com/openshift/origin/blob/v3.11.0/pkg/route/controller/ingress/ingress.go#L576
* Error seems to be present in 4.x: https://github.com/openshift/openshift-controller-manager/blob/release-4.4/pkg/route/ingress/ingress.go#L576 
* Creating an Route with `oc expose service app` creates a correct Route with the name of the service port
* The description of the property `route.spec.port.targetPort` is quite irritating. maybe this is the source of the error. (_The target port on pods selected by the service this route points to. If
     this is a string, it will be looked up as a named port in the target
     endpoints port list._)

**Comment by jgrumboe** [comment]
We can also reproduce the exact issue with OpenShift Master: v3.11.146!

**Comment by ironcladlou** [comment]
Thanks for the detailed report! I'm adding this to our list of bugs to triage for the upcoming 4.3 release, so we should get around to it soon.

**Comment by danehans** [comment]
PR from @derkoe https://github.com/openshift/origin/pull/23951

**Comment by danehans** [comment]
The route api [docs](https://github.com/openshift/api/blob/master/route/v1/types.go#L102-L105) indicate that port should be a service port reference and not a pod port.

**Comment by Miciah** [comment]
@mhofstetter, can you post the corresponding endpoints resource and the pod spec, and verify that the service is working?  I don't understand how the service can be working if the pod does not have a port named "default".

**Comment by mhofstetter** [comment]
hi @Miciah 

the goal is not that the pod doesn't have a port named `default`. the issue is that the current implementation of the Ingress -> Route reconciliation (which creates OpenShift Routes for each Kubernetes Ingress) expects that the name of the port on the `Service` and on the `Pod` needs to be the same - otherwise the created `Route` isn't valid. and that shouldn't be the case.

**Comment by derkoe** [comment]
This only happens when the Service references the Pod port via name (as in the example above with `default`) and the Service port name is different from the Pod port name (or the Service port name is not set).

So, there are tow workarounds:
1. Use the numeric port value of the Pod port in the Services's `targetPort`
2. Use the same name in the Service and Pod port spec

The fix in https://github.com/openshift/openshift-controller-manager/pull/35 and #23951 is still not complete - this is more complex than I thought because the Route's `targetPort` semantics are not very clear.

**Comment by Miciah** [comment]
https://github.com/openshift/openshift-controller-manager/pull/53 fixes the issue in 4.4, so I believe we can close this issue.
/close

**Comment by jgrumboe** [comment]
@Miciah thx, good to hear that the issue is fixed. When will this bugfix be available in 3.11?

**Comment by jgrumboe** [comment]
For everyone searching. Bug should be also fixed in 3.11.232
https://access.redhat.com/errata/RHBA-2020:2477
